site/js/modules/webui/input.js

```diff
@@ -0,0 +1,187 @@
+import * as Autograder from '../autograder/base.js';
+
+// A general representation of a user input field.
+class Field {
+    constructor(
+            name, displayName,
+            {
+                type = 'text', underlyingType = 'string', required = false, placeholder = '',
+                inputClass = '', attributes = '', labelBefore = true, extractInputFunc = undefined
+            } = {}) {
+        // The name of the field which will be used for targeting the field.
+        this.name = name;
+
+        // The display name that will be shown to the user.
+        this.displayName = displayName;
+
+        // The HTML input type.
+        // Supports checkbox, email, number, password, and text types.
+        this.type = type;
+
+        // The underlying type of the field.
+        // Non-standard types are parsed to JSON.
+        // If a non-standard type cannot be parsed to JSON,
+        // a validation error is raised.
+        this.underlyingType = underlyingType;
+
+        // Flags the field requires user input.
+        this.required = required;
+
+        // The placeholder text for the input.
+        this.placeholder = placeholder;
+
+        // Optional classes that are attached to the input.
+        this.inputClass = inputClass;
+
+        // Any additional attributes to the input field.
+        // If the field is required, the required attribute will be added automatically.
+        this.attributes = attributes;
+
+        // Determines the position of the HTML label with respect to the input.
+        this.labelBefore = labelBefore;
+
+        // A custom function for extracting the value from an input.
+        // If a default value extraction cannot be inferred, the value is parsed using JSON.
+        this.extractInputFunc = extractInputFunc;
+    }
+
+    validate() {
+        if ((this.name == undefined) || (this.name == '')) {
+            console.error(`Input field cannot have an empty name: ${JSON.stringify(this)}.`);
+        }
+
+        if ((this.displayName == undefined) || (this.displayName == '')) {
+            console.error(`Input field cannot have an empty display name: ${JSON.stringify(this)}.`);
+        }
+    }
+
+    toHTML() {
+        this.validate();
+
+        let attributes = this.attributes;
+        let displayName = this.displayName;
+        if ((this.required) && (this.placeholder === "")) {
+            attributes += ' required';
+            displayName += ` <span class="required-color">*</span>`;
+        }
+
+        let inputFieldHTML = [
+            `<label for="${this.name}">${displayName}</label>`,
+            `<input type="${this.type}" id="${this.name}" name="${this.name}" placeholder="${this.placeholder}" ${attributes}/>`,
```

Comment:
What about drop downs?

As commented above, this will probably need to be expanded out to a system where we know how to handle specific types (which are currently referred to as "underlyingTypes") and make the correct input element (which may not always be an input tag) for it.